### PR TITLE
Conversion to monorepo

### PR DESCRIPTION
**Pull request by jkaster**
Lots of build/deploy improvements including adding lerna

There are now several npm `packages` in this repo:
- `scripts` has the specification fetching, swagger to OAS conversion, and SDK language source file generation
- `sdk` is the Typescript/Javascript SDK
- `sdk-codegen` is the browser-compatible package for processing OpenAPI specifications and language SDK output to strings
- `sdk-utils` is a very simple core package used by both `scripts` and `sdk-codegen`

`yarn test` is still broken but this is greatly improved from what was available before.

There's probably still lots of tweaks left to do for this branch but it will be good to start the conversation about it now

### CODE REVIEW

File: yarn-error.log [hunk 1]
```diff
@@ -0,0 +1,10373 @@
+Arguments:
```
Comment (josephaxisa) [inline yarn-error.log:1]: Did you check this in intentionally?

File: packages/sdk-codegen/src/sdkModels.ts [hunk 2]
```diff
@@ -1338,7 +1338,7 @@ export class ApiModel implements ISymbolTable, IApiModel {
   }
 
   get version(): string {
-    return this.schema?.version || ''
+    return this.schema?.info.version || ''
```
Comment (bryans99) [inline packages/sdk-codegen/src/sdkModels.ts:1341]: I think I ran into this 😄

File: packages/sdk/README.md [hunk 3]
```diff
@@ -54,7 +54,7 @@ When the SDK is installed and the server location and API credentials are config
 Verify authentication works and that API calls will succeed with code similar to the following:
 
 ```typescript
-import { LookerNodeSDK } from '@looker/sdk'
+import { LookerNodeSDK } from ''
```
Comment (bryans99) [inline packages/sdk/README.md:57]: We no longer import from the @looker/sdk? Why is this the case now that we have packages?

File: packages/sdk/README.md [hunk 4]
```diff
@@ -113,7 +113,7 @@ for example, which will generate files to
 Code similar to the following can be used to develop with both the 3.1 and 4.0 SDKs in the same source file:
 
 ```typescript
-import { Looker40SDK, Looker31SDK, NodeSession, NodeSettingsIniFile } from '@looker/sdk'
+import { Looker40SDK, Looker31SDK, NodeSession, NodeSettingsIniFile } from ''
```
Comment (bryans99) [inline packages/sdk/README.md:116]: Same question as previous

File: packages/sdk/src/rtl/authSession.ts [hunk 5]
```diff
@@ -163,7 +162,7 @@ export abstract class AuthSession implements IAuthSession {
       return result.value
     } else {
       if (result instanceof Buffer) {
-        throw sdkError({message: result.toString(utf8)})
+        throw sdkError({message: result.toString('utf8')})
```
Comment (bryans99) [inline packages/sdk/src/rtl/authSession.ts:165]: I see constants for 'utf8' elsewhere. Why the change?

File: packages/sdk/src/rtl/transport.js [hunk 6]
```diff
@@ -0,0 +1,320 @@
+"use strict";
```
Comment (bryans99) [inline packages/sdk/src/rtl/transport.js:1]: As per discussion. Should, probably not be checked in

File: package.json [hunk 7]
```diff
@@ -1,76 +1,95 @@
...
-        "ts-jest": "^24.1.0",
-        "ts-node": "^8.1.0",
-        "typescript": "^3.6.0"
-    }
+  "private": true,
+  "workspaces": [
+    "packages/*"
+  ],
+  "name": "sdk-codegen-all",
+  "version": "0.2.0-beta-3",
```
[the remark this answers is not in the document]
Comment (jkaster) [reply package.json:7]: Nah, that's a private package.

File: packages/sdk-codegen/src/sdkModels.ts [hunk 8]
```diff
@@ -24,7 +24,7 @@
 
 import * as OAS from 'openapi3-ts'
 import md5 from 'blueimp-md5'
-import { HttpMethod, ResponseMode, responseMode, StatusCode } from '../typescript/looker/rtl/transport'
+import { HttpMethod, ResponseMode, responseMode, StatusCode } from '@looker/sdk/dist/browser'
```
[the remark this answers is not in the document]
Comment (jkaster) [reply packages/sdk-codegen/src/sdkModels.ts:27]: Yeah, working on removing the `/dist` bit

File: packages/sdk-codegen/src/sdkModels.ts (hunk 2 above)
Comment (jkaster) [reply packages/sdk-codegen/src/sdkModels.ts:1341]: So did @josephaxisa

File: packages/sdk/README.md (hunk 3 above)
Comment (jkaster) [reply packages/sdk/README.md:57]: IntelliJ messed up a rename refactor. Thanks, I'll fix it.

File: packages/sdk/README.md (hunk 4 above)
Comment (jkaster) [reply packages/sdk/README.md:116]: same answer. Thanks for noticing.

File: packages/sdk/src/rtl/authSession.ts (hunk 5 above)
Comment (jkaster) [reply packages/sdk/src/rtl/authSession.ts:165]: importing and cross-package dependency issues. Seemed heavy handed after monorepo work.

File: packages/sdk/src/rtl/transport.js (hunk 6 above)
Comment (jkaster) [reply packages/sdk/src/rtl/transport.js]: thanks, removed.

File: bin/smoke [hunk 9]
```diff
@@ -29,25 +30,26 @@ fi
...
   cd ..
 fi
 
 if [[ $tests == *swift* ]]
 then
-  echo "Smoking Swift SDK ..."
-  cd swift/looker
-  xcodebuild test -project looker.xcodeproj -scheme looker-Package -sdk iphonesimulator -destination 'platform=iOS Simulator,name=iPhone 11,OS=13.4' | xcpretty --test --color
+  cd swift/looker || exit
+  echo "Smoking Swift SDK if xcodebuild stops failing miserably, contradicting the swift package configuration ..."
```
[the remark this answers is not in the document]
Comment (jkaster) [reply bin/smoke:50]: Updated. Also, I can't land this branch until the unit tests are working again. Hopefully that can be cleanly resolved shortly.

File: packages/sdk/src/rtl/authSession.ts (hunk 5 above)
Comment (jkaster) [reply packages/sdk/src/rtl/authSession.ts:165]: Also, this is the only utf8 ref in that source file, so I didn't think a const added value. Happy to refactor this in the future as part of the evolving monorepo work if that's a code smell to you. I think 
```typescript
export const utf8encoding = {encoding: 'utf-8'}
```
might be the const to import in those scenarios in that case

File: bin/smoke [hunk 10]
```diff
@@ -47,7 +47,7 @@ fi
 if [[ $tests == *swift* ]]
 then
   cd swift/looker || exit
-  echo "Smoking Swift SDK if xcodebuild stops failing miserably, contradicting the swift package configuration ..."
+  echo "Smoking Swift SDK (this could be broken right now, not need to report it) ..."
```
Comment (bryans99) [inline bin/smoke:50]: > not need to report it
Should be 
> no need to report it
?

File: packages/sdk/src/browser.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+export * from './rtl/apiMethods'
+export * from './rtl/apiSettings'
+export * from './rtl/authToken'
+export * from './rtl/authSession'
```
Comment (bryans99) [inline packages/sdk/src/browser.ts:28]: You've exclude node (which makes sense).
 Would it also make sense to follow the same pattern with browser and extension?

File: packages/sdk/src/node.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+export * from './rtl/apiMethods'
+export * from './rtl/apiSettings'
+export * from './rtl/authToken'
+export * from './rtl/authSession'
+// export * from './rtl/browserSdk'
+// export * from './rtl/browserSession'
+// export * from './rtl/browserTransport'
+export * from './rtl/constants'
+export * from './rtl/corsSession'
+export * from './rtl/extensionSdk'
```
Comment (bryans99) [inline packages/sdk/src/node.ts:34]: You excluded browser. Suspect you should also exclude extension as it is a variation of browser

File: packages/sdk/src/rtl/authSession.ts (hunk 5 above)
Comment (bryans99) [reply packages/sdk/src/rtl/authSession.ts:165]: I'm okay with this. It was an observation more than anything else

File: yarn-error.log (hunk 1 above)
Comment (jkaster) [reply yarn-error.log]: Thanks, no. Removed.

File: packages/sdk-codegen/src/python.gen.spec.ts [hunk 13]
```diff
@@ -23,7 +23,8 @@
  */
 
 import { PythonGen } from './python.gen'
-import { apiTestModel } from './testUtils'
+// @ts-ignore
```
[the remark this answers is not in the document]
Comment (jkaster) [reply packages/sdk-codegen/src/python.gen.spec.ts:26]: It's ignoring a warning/error about going beyond the boundary of the `rootDir`. The import still needs to be resolved

File: bin/smoke (hunk 10 above)
Comment (jkaster) [reply bin/smoke]: fixed

File: packages/sdk/src/browser.ts (hunk 11 above)
Comment (jkaster) [reply packages/sdk/src/browser.ts:28]: Are you saying the extension package references should be removed from `browser.ts`?

File: packages/sdk/src/node.ts (hunk 12 above)
Comment (jkaster) [reply packages/sdk/src/node.ts]: Done, thanks.

**Comment by jkaster** [comment]
This monorepo package work is finally getting close.
- all packages are building
- publishing of canary builds works
- the code generator works for fetching specifications and generating all SDK methods and types
- all the tests are back to parity with what is currently on master, with all typescript tests passing and the other SDK tests passing as much as they were before
- **New Feature** `rawRequest()` is implemented for the TS/JS SDK
- there are surely still configuration tweaks, build optimizations, lint fixes, readme updates, getting rid of the `dist/` folder in bundled packages,  and more to do, but this branch is *functional*

**Comment by joeldodge79** [comment]
I just checked out your branch: `git pull --rebase origin master` auto-merges so you won't have to manually fix the conflicts

File: packages/sdk/src/rtl/extensionTransport.ts [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+  async rawRequest(
+    method: HttpMethod,
+    path: string,
+    queryParams?: any,
+    body?: any,
+    authenticator?: any,
+    options?: Partial<ITransportSettings>
+  ): Promise<IRawResponse> {
+    options = options ? { ...this.options, ...options } : this.options
```
[the remark this answers is not in the document]
Comment (jkaster) [reply packages/sdk/src/rtl/extensionTransport.ts:53]: Ok. I'll remove the options spread

File: packages/sdk/src/rtl/extensionTransport.ts [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+
+  async request<TSuccess, TError>(
+    method: HttpMethod,
+    path: string,
+    queryParams?: any,
+    body?: any,
+    authenticator?: any,
+    options?: Partial<ITransportSettings>
+  ): Promise<SDKResponse<TSuccess, TError>> {
+    options = options ? { ...this.options, ...options } : this.options
```
[the remark this answers is not in the document]
Comment (jkaster) [reply packages/sdk/src/rtl/extensionTransport.ts]: removed the spread

**Comment by jkaster** [comment]
There's still more to do but that can be in subsequent PRs. This is working well enough for the existing multi-package restructuring.
